Sources/WebSocketKit/WebSocket.swift

```diff
@@ -91,23 +91,36 @@ public final class WebSocket {
 
     public func close(code: WebSocketErrorCode = .goingAway) -> EventLoopFuture<Void> {
         let promise = self.eventLoop.makePromise(of: Void.self)
-        self.close(code: code, promise: promise)
+        self.close(code: code, codeToSend: code, promise: promise)
         return promise.futureResult
     }
 
     public func close(
         code: WebSocketErrorCode = .goingAway,
         promise: EventLoopPromise<Void>?
     ) {
+        self.close(code: code, codeToSend: code, promise: promise)
+    }
+
+    /// This is used for setting a closeCode on the socket without necessarily sending that code to the peer. This is
+    /// because close code 1005 and 1006 are used to report errors to the application, but must never be sent over
+    /// the wire (per https://tools.ietf.org/html/rfc6455#section-7.4)
+    internal func close(
+        code: WebSocketErrorCode,
+        codeToSend: WebSocketErrorCode,
+        promise: EventLoopPromise<Void>? = nil)
+    {
         guard !self.isClosed else {
             promise?.succeed(())
             return
         }
         self.isClosed = true
         self.closeCode = code
-        
+
+        assert(codeToSend != .unknown(1005) && codeToSend != .unknown(1006), "Codes 1005 and 1006 are not allowed to be sent as a status code.")
```

Comment:
Could we simply switch on the `code` here and translate it to `.normalClosure` for unsupported cases? For example:

```swift
let codeToSend: WebSocketErrorCode
switch code {
case .unknown(let code), code == 1005 || code == 1006:
    codeToSend == .normalClosure // or .goingAway ?
default:
    codeToSend == code
}

self.closeCode = code
buffer.write(webSocketErrorCode: codeToSend)
```

Reply:
Your suggestion makes way more sense than the way I did it. Thank you!

I've updated the PR. I changed the method of checking the code a bit, so that it will be resilient when code 1005 and 1006 are officially part of the enum instead of being packed as unknown. https://github.com/apple/swift-nio/pull/1410